crates/ruff/src/rules/refurb/rules/repeated_append.rs

```diff
@@ -0,0 +1,377 @@
+    // given statement, find its position there, and match all 'appends' from there.
+    let siblings: &'a [Stmt] = if checker.semantic().at_top_level() {
+        // If the statement is at the top level, we should go to the parent module.
+        // Module is available in the definitions list.
+        let Definition::Module(ref module_def) =
+            checker.semantic().definitions[DefinitionId::module()]
+        else {
+            return None;
+        };
+        module_def.python_ast
+    } else {
+        // Otherwise, go to the parent, and take its body as a sequence of siblings.
+        let parent = checker.semantic().current_statement_parent()?;
+        traversal::suite(stmt, parent)?
+    };
+
+    let stmt_index = siblings.iter().position(|x| x == stmt)?;
+
+    // We shouldn't repeat the same work for many 'appends' that go in a row. Let's check
+    // that this statement is at the beginning of such a group.
+    if stmt_index != 0 && match_append(checker, &siblings[stmt_index - 1]).is_some() {
+        return None;
+    }
+
+    // Starting from the current statement, let's match all appends and make a vector.
+    Some(
+        siblings
+            .iter()
+            .skip(stmt_index)
+            .map_while(|next| match_append(checker, next))
+            .collect(),
+    )
+}
+
+fn group_appends<'a>(appends: &[Append<'a>]) -> Vec<AppendGroup<'a>> {
+    // We want to go over the given list of appends and group the by receivers.
+    let mut map: HashMap<NodeId, AppendGroup<'a>> = HashMap::new();
+    let mut last_binding = {
+        let group = get_or_add(&mut map, appends.first().unwrap());
+        get_binding_stmt(group)
+    };
+
+    for append in appends.iter().skip(1) {
+        let group = get_or_add(&mut map, append);
+        let binding = get_binding_stmt(group);
+        if binding != last_binding {
+            // If the group is not brand new, and the previous group was different,
+            // we should mark it as "non-consecutive".
+            if group.appends.len() > 1 {
+                group.is_consecutive = false;
+            }
+            last_binding = binding;
+        }
+    }
+
+    map.into_values()
+        // Groups with just one element are not interesting for us.
+        .filter(|group| group.appends.len() > 1)
+        .collect()
+}
+
+#[inline]
+fn get_binding_stmt(group: &AppendGroup) -> NodeId {
+    group.appends.first().unwrap().binding.source.unwrap()
+}
+
+#[inline]
+fn get_or_add<'a, 'b>(
+    map: &'b mut HashMap<NodeId, AppendGroup<'a>>,
+    append: &Append<'a>,
+) -> &'b mut AppendGroup<'a> {
+    let binding_stmt = append.binding.source.unwrap();
+    let group = map.entry(binding_stmt).or_insert(AppendGroup {
+        appends: vec![],
+        is_consecutive: true,
+    });
+    group.appends.push(append.clone());
+    group
+}
+
+fn make_suggestion(group: &AppendGroup, generator: Generator) -> String {
```

Comment:
Can we add rustdoc for some of these more involved methods? Just to outline its purpose.

Reply:
Sure, I just didn't know what's the official policy on private functions and documentation.